### PR TITLE
[JUJU-3447] Enable ModelFirewaller for OpenStack

### PR DESCRIPTION
**Pull request by jack-w-shaw**
This work has already been done for aws. Add OpenStack to the list of
providers which support this
https://github.com/juju/juju/pull/15329

As with aws, model ssh ingress is now managed by `ssh-allow` model config item.
Also, api-port is now only opened on the controller model 

This was mostly trivial, and involved removing a lot of responsibilities
from the OpenStack provider

This has involved updating the go-goose dependency

Also as a flyby fix an bug in config. ssh-allow was defaulting to the
wrong value if the entry was absent. Write a test for this case as well

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing
- [x] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

### Verify ssh-allow

Deploy a controller to opensatck, add a model and deploy a machine
```
$ juju bootstrap openstack stack
$ juju add-model m
$ juju add-machine
```

Then find the model-level security group name, put it in an envvar. For me:
```
$ export MODEL_GROUP=juju-83a82f66-bfd4-4bb0-812a-e3edfb67d64d-57d8be23-4f99-4de0-8946-bb9e89d21f78
```

And now verify ssh-allow default settings
```
$ juju model-config ssh-allow
0.0.0.0/0,::/0
$ openstack security group show $MODEL_GROUP -f json | jq -r '.rules[] | select(.port_range_min == 22) | .remote_ip_prefix'
0.0.0.0/0
::/0
```

Now configure ssh-allow
```
$ juju model-config ssh-allow="192.168.0.0/24"
$ openstack security group show $MODEL_GROUP -f json | jq -r '.rules[] | select(.port_range_min == 22) | .remote_ip_prefix'
192.168.0.0/24
```

Add a machine to verify this doesn't reset things
```
$ juju model-config ssh-allow="192.168.0.0/24"
$ openstack security group show $MODEL_GROUP -f json | jq -r '.rules[] | select(.port_range_min == 22) | .remote_ip_prefix'
192.168.0.0/24
```

Add a new model pre-configured to check the config comes into effect even if no machines are present when it's set. (You will need to reset `MODEL_GROUP`
```
$ juju add-model m2 --config ssh-allow="192.168.2.0/24"
$ juju add-machine
(wait)
$ openstack security group show $MODEL_GROUP -f json | jq -r '.rules[] | select(.port_range_min == 22) | .remote_ip_prefix'
192.168.0.0/24
```

### Verify multiple machines can be added to the same model
```
$ juju add-model m
$ juju add-machine
(wait)
$ juju add-machine
(wait)
$ juju status
Model  Controller  Cloud/Region           Version      SLA          Timestamp
m      mstack      microstack/microstack  3.2-beta4.1  unsupported  14:13:42+01:00

Machine  State    Address         Inst id                               Base          AZ    Message
0        started  192.168.222.61  9b7c14e8-dd6f-4cb3-ad19-0ffa8e852f48  ubuntu@22.04  nova  ACTIVE
1        started  192.168.222.54  5a51c5d4-af57-4231-b3d9-afc28736467a  ubuntu@22.04  nova  ACTIVE
```

### Verify `juju expose` still exposes
```
$ juju deploy postgresql
(wait)
$ juju expose postgresql
$ openstack security group show $MACHINE_GROUP
+-----------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Field           | Value                                                                                                                                                                                                                                              |
+-----------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| created_at      | 2023-04-25T13:14:34Z                                                                                                                                                                                                                               |
| description     | juju group                                                                                                                                                                                                                                         |
| id              | 1069e8a6-a3b1-4aee-8471-d02f48d395e1                                                                                                                                                                                                               |
| location        | cloud='', project.domain_id=, project.domain_name='default', project.id='8910e2a31e8a4834a266bfd5dc99d80c', project.name='admin', region_name='', zone=                                                                                            |
| name            | juju-20810b78-b20d-4543-837e-9715be4fcda1-01de2e1c-76a3-40a0-864f-694e16e2164b-2                                                                                                                                                                   |
| project_id      | 8910e2a31e8a4834a266bfd5dc99d80c                                                                                                                                                                                                                   |
| revision_number | 3                                                                                                                                                                                                                                                  |
| rules           | created_at='2023-04-25T13:14:34Z', direction='egress', ethertype='IPv6', id='11996358-35f0-4316-90f4-9eef38a623fc', updated_at='2023-04-25T13:14:34Z'                                                                                              |
|                 | created_at='2023-04-25T13:14:34Z', direction='egress', ethertype='IPv4', id='22a74304-8426-42cf-b93b-2a8a8b85357e', updated_at='2023-04-25T13:14:34Z'                                                                                              |
|                 | created_at='2023-04-25T13:18:28Z', direction='ingress', ethertype='IPv4', id='a847c482-a5a7-4935-af77-dcd52267a652', port_range_max='5432', port_range_min='5432', protocol='tcp', remote_ip_prefix='0.0.0.0/0', updated_at='2023-04-25T13:18:28Z' |
|                 | created_at='2023-04-25T13:18:28Z', direction='ingress', ethertype='IPv6', id='ae88d48b-53ff-4cbb-94a4-9fafa9ae4e9a', port_range_max='5432', port_range_min='5432', protocol='tcp', remote_ip_prefix='::/0', updated_at='2023-04-25T13:18:28Z'      |
| stateful        | True                                                                                                                                                                                                                                               |
| tags            | []                                                                                                                                                                                                                                                 |
| updated_at      | 2023-04-25T13:18:28Z                                                                                                                                                                                                                               |
+-----------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### Verify api-port
```
$ juju bootstrap openstack openstack --config api-port=17777
(wait)
$ openstack security group show $MODEL_GROUP -f json | jq -r '.rules[] | select(.port_range_min == 17777) | .remote_ip_prefix'
0.0.0.0/0
```

### Verify autocert-dns-name
```
$ juju bootstrap openstack openstack --config autocert-dns-name=example.com
(wait)
$ openstack security group show $MODEL_GROUP -f json | jq -r '.rules[] | select(.port_range_min == 80) | .remote_ip_prefix'
0.0.0.0/0
```

### CODE REVIEW

File: provider/openstack/firewaller.go [hunk 1]
```diff
@@ -397,111 +343,82 @@ func (c *neutronFirewaller) ensureGroup(name string, rules []neutron.RuleInfoV2)
...
-	// Find rules we want to delete, that we have but don't want, and
-	// delete them.
-	// Define a removal set to ensure that we only ever delete a ruleID once.
-	remove := set.NewStrings()
-	for k := range have {
-		// Neutron creates 2 egress rules with any new Security Group.
-		// Keep them.
-		if _, ok := want[k]; !ok && k.Direction != "egress" {
-			remove.Add(have[k])
+	if isModelGroup {
```
Comment (hmlanigan) [inline provider/openstack/firewaller.go:346]:
```
if !isModelGroup {
    return group, nil
}
```
Reduces indentation and helps with readability.

**Review by hmlanigan — COMMENTED** [review]
Why are there no changes to machine security group rules when an application is exposed?

Will continue when the QA steps are updated.

File: provider/openstack/firewaller.go [hunk 2]
```diff
@@ -397,111 +343,82 @@ func (c *neutronFirewaller) ensureGroup(name string, rules []neutron.RuleInfoV2)
...
-	add := make(ruleInfoSet)
-	for k := range want {
-		if _, ok := have[k]; !ok {
-			add[k] = want[k]
-		}
+	return group, nil
+}
+
+func (c *neutronFirewaller) ensureInternalRules(neutronClient *neutron.Client, group neutron.SecurityGroupV2) error {
+	rules := []neutron.RuleInfoV2{
```
[the remark this answers is not in the document]
Comment (jack-w-shaw) [reply provider/openstack/firewaller.go:367]: The ones that were dropped were the rules opening ports 22 and the apiPort to the world. Now that ModelFirewaller is implemented, these are no longer the responsibility of the provider and will be handled by the firewaller worker

**Comment by jack-w-shaw** [comment]
> Why are there no changes to machine security group rules when an application is exposed?

I have just tested this. It still works on my machine. I have added this as a QA step

This is ready for another review now, however to get QA working the version in version/version.go will need to be bumped. Either wait for me to merge https://github.com/juju/juju/pull/15533 and rebase, or bump the version yoruself

**Review by hmlanigan — CHANGES_REQUESTED** [review]
QA: By default, new models do not get port 22 opened, once you set or reset the ssh-allow model config the security group value is set. This log message was found:
```controller-0: 14:34:03 WARNING juju.worker.firewaller Model firewall not found so cannot re-write firewall rule ["22/tcp from 0.0.0.0/0,::/0"]. Ignoring```

Code looks good, just the 1 bug in qa to fix.

**Pull request closed without merging**